File: 648.cpp
struct Solution {
    struct TrieNode {
        int len = 0;
        int next[26] = {0};
    };

    string replaceWords(vector<string>& dictionary, string &sentence) {
        vector<TrieNode> trie(1);
        for (string &w : dictionary){
            int ptr = 0;
            for (char &c : w){
                if (trie[ptr].next[c - 'a'] == 0){
                    trie[ptr].next[c - 'a'] = trie.size();
                    trie.emplace_back();
                }
                ptr = trie[ptr].next[c - 'a'];
            }
            trie[ptr].len = w.size();
        }
        string res;
        int idx = 0;
        while (idx < sentence.size()){
            int nextIdx = sentence.find(' ', idx);
            if (nextIdx == string::npos){
                nextIdx = sentence.size();
            }
            int ptr = 0, repLen = nextIdx - idx;
            for (int i = idx; i < nextIdx; ++i){
                char &c = sentence[i];
                if (trie[ptr].next[c - 'a'] == 0){
                    break;
                }
                ptr = trie[ptr].next[c - 'a'];
                if (trie[ptr].len != 0){
                    repLen = trie[ptr].len;
                    break;
                }
            }
            res += sentence.substr(idx, repLen);
            res += " ";
            idx = nextIdx + 1;
        }
        res.pop_back();
        return res;
    }
};

File: 648.py
class Solution:
    def replaceWords(self, dictionary: List[str], sentence: str) -> str:
        trie = dict()
        for w in dictionary:
            ptr = trie
            for c in w:
                if c not in ptr:
                    ptr[c] = dict()
                ptr = ptr[c]
            ptr['='] = w
        def rep(w: str) -> str:
            ptr = trie
            for c in w:
                if c not in ptr:
                    return w
                ptr = ptr[c]
                if '=' in ptr:
                    return ptr['=']
            return w
        return " ".join(rep(w) for w in sentence.split())
